findingTheBoundaryWithBinarySearch.js

```diff
@@ -0,0 +1,70 @@
+/*
+As similar to the previous problem, we're going to approach this problem using a binary search algorithm and 
+divide a porition of the list in half, until we've narrow down the location of the target.
+
+Since we're trying to find the first sign of the true boolean, we're going to set two pointers. One pointer
+will be at the start and the other at the end. If a left pointer reaches a false boolean, we'll
+know our target is more to the right. If a right pointer reaches a true boolean, there's a
+possibilty that it's not the earlist one, so we'll continue searching one until we 
+reach a true boolean next to a false one and then return the current index of the
+true boolean.
+*/
+
+/*
+Question:
+An array of boolean values is divided into two sections; the left section 
+consists of all false and the right section consists of all true. Find the 
+First True in a Sorted Boolean Array of the right section, i.e. the index 
+of the first true element. If there is no true element, return -1.
+*/
+
+// TODO: write your code here
+
+const NOT_FOUND = 1;
+
+const getMidIdx = (leftIdx, rightIdx) => {
+    const idxRange = rightIdx - leftIdx;
+    const midIdx = leftIdx + Math.floor(idxRange / 2);
+
+    return midIdx;
+};
+
+const findBoundary = (booleans) => {
+    let leftIdx = 0;
+    let rightIdx = booleans.length - 1;
+
+    let boundaryIdx = NOT_FOUND;
+
+    while (leftIdx <= rightIdx) {
+        const midIdx = getMidIdx(leftIdx, rightIdx);
+        console.log(midIdx);
+        const midBoolean = booleans[midIdx];
+
+        if (midBoolean) {
```

Comment:
P: Why are you updating your indices in this way? Explain the logic behind these ~5 lines.

Reply:
It's because we're trying to find the first true boolean that appears in the array and if we happen to reach a true boolean, there's a possibility that it's not the first one, so we move the right pointer to the left to narrow down the search. There's no reason to increase it because we're only concerned about the first true boolean, which would be more to the left.

If we reach a false boolean, we move our left pointer to the right to narrow down the search, ignoring everything else from the left side and continue looking in the right path.

Reply:
Beautiful explanation!

Just one detail I'd add: when we find a true, it is POSSIBLE that we have currently found the correct answer. So, we temporarily store the current index as the "best answer so far", and then search the left chunk to see if there is a "better answer".

Reply:
Yes, I did forget to mention that. Good point.